src/agent/Core/ApplicationPool/Group/ProcessListManagement.cpp

```diff
@@ -64,71 +64,126 @@ Group::findProcessWithStickySessionId(unsigned int id) const {
+				(process->generation == fallbackProcess->generation && process->spawnStartTime == fallbackProcess->spawnStartTime && process->busyness() < fallbackProcess->busyness())
+			) {
+				fallbackProcess = process;
+			}
+		} else if (!process->isTotallyBusy() && (bestProcess == nullptr ||
+				   process->generation > bestProcess->generation ||
+				   (process->generation == bestProcess->generation && process->spawnStartTime < bestProcess->spawnStartTime) ||
+				   (process->generation == bestProcess->generation && process->spawnStartTime == bestProcess->spawnStartTime && process->busyness() < bestProcess->busyness())
+		)) {
+			bestProcess = process;
 		}
 	}
-
-	if (leastBusyProcessIndex == -1) {
-		return NULL;
-	} else {
-		return enabledProcesses[leastBusyProcessIndex].get();
+	if (bestProcess == nullptr) {
+		return fallbackProcess;
 	}
+	return bestProcess;
 }
 
 Process *
-Group::findProcessWithLowestBusyness(const ProcessList &processes) const {
+Group::findBestProcess(const ProcessList &processes) const {
 	if (processes.empty()) {
-		return NULL;
+		return nullptr;
 	}
 
-	int lowestBusyness = -1;
-	Process *leastBusyProcess = NULL;
+	Process *bestProcess = nullptr;
+	Process *fallbackProcess = nullptr;
 	ProcessList::const_iterator it;
 	ProcessList::const_iterator end = processes.end();
 	for (it = processes.begin(); it != end; it++) {
 		Process *process = (*it).get();
-		int busyness = process->busyness();
-		if (lowestBusyness == -1 || lowestBusyness > busyness) {
-			lowestBusyness = busyness;
-			leastBusyProcess = process;
+
+	 if (process->isTotallyBusy() && bestProcess == nullptr) {
+			if (fallbackProcess == nullptr ||
+				process->generation > fallbackProcess->generation ||
+				(process->generation == fallbackProcess->generation && process->spawnStartTime < fallbackProcess->spawnStartTime) ||
+				(process->generation == fallbackProcess->generation && process->spawnStartTime == fallbackProcess->spawnStartTime && process->busyness() < fallbackProcess->busyness())
+			) {
+				fallbackProcess = process;
+			}
+	 } else if (!process->isTotallyBusy() && (bestProcess == nullptr ||
+			process->generation > bestProcess->generation ||
+			(process->generation == bestProcess->generation && process->spawnStartTime < bestProcess->spawnStartTime) ||
+			(process->generation == bestProcess->generation && process->spawnStartTime == bestProcess->spawnStartTime && process->busyness() < bestProcess->busyness())
+		)) {
+			bestProcess = process;
 		}
 	}
-	return leastBusyProcess;
+	if (bestProcess == nullptr) {
+		return fallbackProcess;
+	}
+	return bestProcess;
 }
 
 /**
- * Cache-optimized version of findProcessWithLowestBusyness() for the common case.
+ * Cache-optimized version of findBestProcess() for the common case.
  */
 Process *
-Group::findEnabledProcessWithLowestBusyness() const {
+Group::findBestEnabledProcess() const {
```

Comment:
`findBestEnabledProcess()` was originally meant to be optimized for speed by using cache-friendly data structures. The changed version is not really cache-friendly because it chases many pointers (it reads `process->XXX` rather than only from `enabledProcessBusynessLevels`.

I think that's fine though. At this point we don't need to win benchmark contests anymore, so we should prioritize code maintainability. As such, I think we can get rid of `findBestEnabledProcess()` and only use `findBestProcess()`.